internal/provider/resource_ssh_key.go

```diff
@@ -5,183 +5,309 @@ import (
+	resp.Diagnostics.Append(req.Plan.Get(ctx, &data)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+	modelName := data.ModelName.ValueString()
+	modelUUID, err := s.client.Models.ResolveModelUUID(modelName)
 	if err != nil {
-		return diag.FromErr(err)
+		resp.Diagnostics.AddError("Client Error", fmt.Sprintf("Unable to resolve model UUID, got error: %s", err))
+		return
 	}
-
-	payload := d.Get("payload").(string)
-
+	payload := data.Payload.ValueString()
 	user := utils.GetUserFromSSHKey(payload)
 	if user == "" {
-		return diag.Errorf("malformed SSH key, user not found")
+		resp.Diagnostics.AddError("Client Error", "malformed SSH key, user not found")
+		return
 	}
 
-	err = client.SSHKeys.CreateSSHKey(&juju.CreateSSHKeyInput{
+	err = s.client.SSHKeys.CreateSSHKey(&juju.CreateSSHKeyInput{
 		ModelName: modelName,
 		ModelUUID: modelUUID,
 		Payload:   payload,
 	})
-
 	if err != nil {
-		diags = append(diags, diag.FromErr(err)...)
+		resp.Diagnostics.AddError("Client Error", fmt.Sprintf("Unable to create ssh_key, got error %s", err))
+		return
 	}
+	tflog.Trace(ctx, fmt.Sprintf("create ssh_key with payload: %q", payload))
 
-	d.SetId(fmt.Sprintf("sshkey:%s:%s", modelName, user))
+	data.ID = types.StringValue(newSSHKeyID(modelName, user))
+	resp.Diagnostics.Append(resp.State.Set(ctx, &data)...)
+}
 
-	return diags
+func newSSHKeyID(modelName string, user string) string {
+	return fmt.Sprintf("sshkey:%s:%s", modelName, user)
 }
 
-func sshKeyRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
-	client := meta.(*juju.Client)
+func retrieveSSHKeyInfoFromID(id string, d *frameworkdiag.Diagnostics) (string, string) {
+	tokens := strings.Split(id, ":")
+	//If importing with an incorrect ID we need to catch and provide a user-friendly error
+	if len(tokens) != 3 {
+		d.AddError("Malformed ID", fmt.Sprintf("unable to parse model name and user from provided ID: %q", id))
+		return "", ""
+	}
+	return tokens[1], tokens[2]
+}
 
-	// sshkey:model:user
-	tokens := strings.Split(d.Id(), ":")
-	modelName := tokens[1]
-	user := tokens[2]
+func (s *sshKeyResource) Read(ctx context.Context, req resource.ReadRequest, resp *resource.ReadResponse) {
+	// Prevent panic if the provider has not been configured.
+	if s.client == nil {
+		addClientNotConfiguredError(&resp.Diagnostics, "ssh_key", "read")
+		return
+	}
 
-	modelUUID, err := client.Models.GetModelByName(modelName)
-	if err != nil {
-		return diag.FromErr(err)
+	var plan sshKeyResourceModel
+
+	// Read Terraform configuration from the request into the model
+	resp.Diagnostics.Append(req.State.Get(ctx, &plan)...)
+	if resp.Diagnostics.HasError() {
+		return
 	}
 
-	result, err := client.SSHKeys.ReadSSHKey(&juju.ReadSSHKeyInput{
-		ModelName: modelUUID.Name,
-		ModelUUID: modelUUID.UUID,
-		User:      user,
-	})
-	if err != nil {
-		return diag.FromErr(err)
+	modelName, user := retrieveSSHKeyInfoFromID(plan.ID.ValueString(), &resp.Diagnostics)
```

Comment:
Suggest s/retrieveSSHKeyInfoFromID/retrieveModelUserNameFromID/

Reply:
Yeah that's better. Though we better not have another resource that keeps just the model name and user in the ID :)